### PR TITLE
IL-114: Add file system and file encryption

### PR DESCRIPTION
**Pull request by Nuni88**
- AsyncStorage has been replaced with with files stored in a directory.
- All actions that modify cards and messages now use the new file system.
- Cards and messages are now encrypted when written and decrypted when read.
- A key and initialization vector is generated on starting the app for AES encryption. This key is stored in the device's secure keychain/keystore for future use.
- Unused functions such as Delete and Update card have been removed.

Testing procedure:

1. Clear all data.
2. Reload and observe encrypted card and message data in console log.
3. Through wallet, go to add a card. Observe encrypted card data in console log when card is added.
4. Create a new message. Observe encrypted message data in console log when message is created.
5. Perform any other actions that access messages and cards and ensure that the forward-facing behavior is visibly unchanged.

### CODE REVIEW

**Comment by joewetton** [comment]
Can't seem to get cards to load nor add a new card. The error is when I try to load the wallet.
img_051efc666b66-1

**Comment by Nuni88** [comment]
@joewetton Did you make sure to clear all data and reload the app first? And after clearing all data and reloading, if you look at your console log, do you see the encrypted cards and message data? It might also be useful if you could add console log lines in init after key and iv are set to get the values of them, because it might be a problem with them not being set properly.

**Comment by joewetton** [comment]
When I try to add a card.. seems the messages.dat and cards.dat files aren't getting made.
screen shot 2018-05-12 at 12 16 25 pm

**Comment by joewetton** [comment]
https://github.com/wkh237/react-native-fetch-blob/wiki/File-System-Access-API#encodingstring

**Comment by Nuni88** [comment]
I just saw that too. I'll try and get a fix going here for iOS.

**Comment by bnelson777** [comment]
Just tested it too for ios, same issue. Looks like you're working on a fix though. 💯

**Comment by Nuni88** [comment]
@joewetton @bnelson777 I pushed up some changes. I don't _think_ they'll fix the iOS issue, but if one of you could test to find out, that would be helpful.

**Comment by bnelson777** [comment]
@Nuni88 Sure testing now

**Comment by bnelson777** [comment]
Tried again with same issues. Seems the unhandled promise rejection error is kinda sporadic for me (showing up at random screens sometimes and not others). I tried running it on my device twice. I got the the below warning the first time but not the second time, so not sure if that helps at all..

img_4557

**Comment by Nuni88** [comment]
@bnelson777 I'm not sure if that warning is related or not. Googling it wasn't too helpful. This is going to be very difficult to troubleshoot without access to an iOS platform, though. If an iOS person is on campus Monday or Tuesday and has some time, maybe we could meet up to troubleshoot the iOS side?

**Comment by bnelson777** [comment]
@Nuni88 Sure I can meet up Monday, I have class in the morning but pretty much free otherwise

**Comment by Nuni88** [comment]
@bnelson777 Cool. I've got class from 2-4, but any time outside of that should be fine.

**Comment by joewetton** [comment]
Narrowed down the problem I think, seems to be an issue with iv and key not being set via sensitive info.

File: app/index.js [hunk 1]
```diff
@@ -35,24 +38,99 @@ import * as ReduxActions from './actions'; //Import your actions
...
         var _this = this;
         //Check if any card data exists
-        AsyncStorage.getItem('carddata', (err, carddata) => {
-            //if it doesn't exist, extract from json file
-            //save the initial data in Async
-            if (carddata === null){
-                AsyncStorage.setItem('carddata', JSON.stringify(CardData.card));
+        RNFetchBlob.fs.readFile(paths.cardsPath, 'utf8')
+        .then((carddata) => {
+            if (carddata === ''){
```
Comment (joewetton) [inline app/index.js:115]: 2018-05-12 15:16:31.956 [info][tid:com.facebook.react.JavaScript] 'key: ', undefined
2018-05-12 15:16:31.956569-0700 idly[1559:30315] 'key: ', undefined
2018-05-12 15:16:31.957 [info][tid:com.facebook.react.JavaScript] 'iv: ', undefined
2018-05-12 15:16:31.957123-0700 idly[1559:30315] 'iv: ', undefined
JSON.stringify(CardData.card) prints the cards

File: app/index.js [hunk 2]
```diff
@@ -35,24 +38,99 @@ import * as ReduxActions from './actions'; //Import your actions
...
+        .catch((err) => {});
+
+        RNFetchBlob.fs.createFile(
+            paths.messagesPath,
+            '',
+            'utf8'
+        )
+        .catch((err) => {});
+
+        SInfo.getItem('key', {})
```
Comment (joewetton) [inline app/index.js:70]: Not getting set on IOS

File: app/index.js [hunk 3]
```diff
@@ -35,24 +38,99 @@ import * as ReduxActions from './actions'; //Import your actions
...
+            else
+                this.setState({key: value});
+        });
+
+        SInfo.getItem('iv', {})
+        .then((value) => {
+            if (value === null){
+                for(var iv = ''; iv.length < 16;)
+                    iv += Math.random().toString(36).substr(2, 1)
+                SInfo.setItem('iv', iv, {});
```
Comment (joewetton) [inline app/index.js]: Diddo.

**Comment by Nuni88** [comment]
@joewetton @bnelson777 One of you want to try testing again? It looks like iOS uses undefined instead of null for keys not present, so I tried changing === null to  == null when making keys which _might_ fix it since == null catches both.

**Comment by joewetton** [comment]
@Nuni88 seems that fixed the issues in componentDidMount, but now the iv and key are empty in the actions GetCards() call.

screen shot 2018-05-13 at 2 00 39 pm
screen shot 2018-05-13 at 2 00 30 pm

**Comment by Nuni88** [comment]
@joewetton That issue should be fixed now.

**Comment by joewetton** [comment]
yep it's working, except our components don't seem to be automatically updating their state like they use to. When I compose a new message and click done the inbox view isn't updated to reflect the new message.

**Comment by Nuni88** [comment]
Yeah, I think it's just an issue with the component loading and rendering before the action is complete, which might just mean we have to somehow force it to wait for the action to complete first.

**Comment by joewetton** [comment]
A timeout of some kind? I'm pretty sure redux is supposed to do this for us. Like it should rerender the entire app on every component load which is use to do.

**Comment by Nuni88** [comment]
Yeah, I'm not too sure. I think it's something to do with the actions doing so much async work to be able to do the encryption and decryption that it throws the timing off.

**Review by joewetton — CHANGES_REQUESTED** [review]
My only concern left is the redux state doesn't update after hitting clearAll() fix that and I think it's good to merge. We can add the redux refreshing issue as a bug for fixing next sprint.

**Comment by Nuni88** [comment]
@joewetton It should be clearing the redux state with clearAll now. The loading issue when modifying cards and messages should be fixed too. There are unhandled promise warnings when modifying cards or messages now though, so I'll try to get rid of those after class today, but the functionality should all be there now.

**Comment by joewetton** [comment]
Great work.

**Review by joewetton — APPROVED** [review]

**Comment by Nuni88** [comment]
Warnings when modifying cards or messages should be gone now. New system forces the app to wait a brief period after modifying cards/messages before going to another screen to allow it to catch up.

**Review by Vancev — APPROVED** [review]
Everything working for android

**Pull request merged**